segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java

```diff
@@ -482,6 +511,109 @@ public boolean isOffline() {
+                                .collect(Collectors.toList());
+                        if (!attributeIds.isEmpty()) {
+                            // This only makes sense if a core attribute was missing.
+                            return getAndCacheAttributes(segmentMetadata, attributeIds, true, timer)
+                                    .thenComposeAsync(attributes -> {
+                                        // Final attempt - now that we should have the attributes cached.
+                                        return this.durableLog.add(operation, timer.getRemaining());
+                                    }, this.executor);
+                        }
+                    }
+
+                    // Anything else is non-retryable; rethrow.
+                    return Futures.failedFuture(ex);
+                });
+    }
+
+    /**
+     * Gets the values of the given (Core and Extended) Attribute Ids for the given segment.
+     *
+     * @param segmentMetadata The SegmentMetadata for the Segment to retrieve attribute values for.
+     * @param attributeIds    A Collection of AttributeIds to retrieve.
+     * @param cache           If true, any Extended Attribute value that is not present in the SegmentMetadata cache will
+     *                        be added to that (using a conditional updateAttributes() call) before completing.
+     * @param timer           Timer for the operation.
+     * @return A CompletableFuture that, when completed normally, will contain the desired result. If the operation failed,
+     * it will be completed with the appropriate exception. If cache==true and the conditional call to updateAttributes()
+     * could not be completed because of a conflicting update, it will be failed with BadAttributeUpdateException, in which
+     * case a retry is warranted.
+     */
+    private CompletableFuture<Map<UUID, Long>> getAndCacheAttributes(SegmentMetadata segmentMetadata, Collection<UUID> attributeIds, boolean cache, TimeoutTimer timer) {
+        // Collect Core Attributes and Cached Extended Attributes.
+        Map<UUID, Long> result = new HashMap<>();
+        Map<UUID, Long> metadataAttributes = segmentMetadata.getAttributes();
+        ArrayList<UUID> extendedAttributeIds = new ArrayList<>();
+        attributeIds.forEach(attributeId -> {
+            Long v = metadataAttributes.getOrDefault(attributeId, Attributes.NULL_ATTRIBUTE_VALUE);
+            if (v != Attributes.NULL_ATTRIBUTE_VALUE) {
+                result.put(attributeId, v);
+            } else if (!Attributes.isCoreAttribute(attributeId)) {
+                extendedAttributeIds.add(attributeId);
+            }
+        });
+
+        // Collect remaining Extended Attributes.
+        CompletableFuture<Map<UUID, Long>> r = this.attributeIndex
+                .forSegment(segmentMetadata.getId(), timer.getRemaining())
+                .thenComposeAsync(idx -> idx.get(extendedAttributeIds, timer.getRemaining()), this.executor);
```

Comment:
A clarification: the call to `idx.get` here will induce a call to `readAllSinceLastSnapshot`, which performs a read of the snapshot in the attribute segment perhaps unnecessarily in the case we have collected all attributes by Line 583. In the case my interpretation is correct, is it worth checking before reading the snapshot from the attribute segment?

Reply:
`extendedAttributeIds` only contains those attribute ids from `attributeIds` that we were unable to fetch from memory (see the construction above). If this method is called with an empty collection, it will just return, without invoking `readAllSinceLastSnapshot`.

Reply:
You're right, that was pretty straightforward actually.